kvserver: fix slow raft proposal gauge

tbg opened the pull request:
There were three problems.

First, we weren't actually counting all slow proposals on each range,
since we were only updating the counter whenever we'd see a new max. If
the maximum proposal was visited first, we'd only record one; in
general the count was anywhere between one and the correct one

The second, bigger, problem is that we were updating a per-Store gauge
in a per-Replica method. So even if some Replica would record a nonzero
value, some other Replica would in all likelihood immediately clobber
that value with a zero again. So in practice, we'd see a zero here most
of the time unless many ranges had slow proposals.

The third problem is the opposite (but is masked by the first two):
since refreshProposalsLocked is the driver of the updates, we need to be
careful to have called it after slow proposals clear out and before we
quiesce. Otherwise, we would still have slow proposals reflected in the
gauge even though they have already completed and the range they were
on has since quiesced.

All of this is now fixed: we record the proper count of slow proposals
per Replica, and sum that up into the per-Store gauge, like we do for
all other Replica-sourced metrics.

We also prevent ranges from quiescing unless the slow count is zero;
this prevents leaking a count if there were slow proposals but they then
resolved but refreshProposalsLocked hasn't been re-invoked yet to clear
the count. This delays quiescing by up to a few ticks (after slow
proposals resolved), which is around a second and thus doesn't matter.

I verified manually, creating a three-node local cluster, and stopping
two nodes. The gauge kept climbing slowly (printouts minute over minute)

> requests_slow_raft{store="1"} 0
> requests_slow_raft{store="1"} 3
> requests_slow_raft{store="1"} 10
> requests_slow_raft{store="1"} 17

In particular note the monotonicity. Before the patch, we would've seen
out-of-order numbers here, since we'd see essentially a "random" Replica's
count (problem 2), which in itself was also "random" (problem 1).

(Note that the replica circuit breakers fail-fast most proposals before
they enter the replication layer, but the probes are still tracked; this
is why the number isn't increasing faster).

Restarting n2 and n3 let the gauge drop back to zero (10s intervals):

> requests_slow_raft{store="1"} 21
> requests_slow_raft{store="1"} 4
> requests_slow_raft{store="1"} 0

Release note: None

Release justification: none, will merge once no justification necessary

tbg committed:
kvserver: extract struct for calcReplicaMetrics

Release justification: None needed
Release note: None

tbg committed:
kvserver: fix slow raft proposal gauge

There were three problems.

First, we weren't actually counting all slow proposals on each range,
since we were only updating the counter whenever we'd see a new max. If
the maximum proposal was visited first, we'd only record one; in
general the count was anywhere between one and the correct one

The second, bigger, problem is that we were updating a per-Store gauge
in a per-Replica method. So even if some Replica would record a nonzero
value, some other Replica would in all likelihood immediately clobber
that value with a zero again. So in practice, we'd see a zero here most
of the time unless many ranges had slow proposals.

The third problem is the opposite (but is masked by the first two):
since refreshProposalsLocked is the driver of the updates, we need to be
careful to have called it after slow proposals clear out and before we
quiesce. Otherwise, we would still have slow proposals reflected in the
gauge even though they have already completed and the range they were
on has since quiesced.

All of this is now fixed: we record the proper count of slow proposals
per Replica, and sum that up into the per-Store gauge, like we do for
all other Replica-sourced metrics.

We also prevent ranges from quiescing unless the slow count is zero;
this prevents leaking a count if there were slow proposals but they then
resolved but refreshProposalsLocked hasn't been re-invoked yet to clear
the count. This delays quiescing by up to a few ticks (after slow
proposals resolved), which is around a second and thus doesn't matter.

I verified manually, creating a three-node local cluster, and stopping
two nodes. The gauge kept climbing slowly (printouts minute over minute)

> requests_slow_raft{store="1"} 0
> requests_slow_raft{store="1"} 3
> requests_slow_raft{store="1"} 10
> requests_slow_raft{store="1"} 17

In particular note the monotonicity. Before the patch, we would've seen
out-of-order numbers here, since we'd see essentially a "random" Replica's
count (problem 2), which in itself was also "random" (problem 1).

(Note that the replica circuit breakers fail-fast most proposals before
they enter the replication layer, but the probes are still tracked; this
is why the number isn't increasing faster).

Restarting n2 and n3 let the gauge drop back to zero (10s intervals):

> requests_slow_raft{store="1"} 21
> requests_slow_raft{store="1"} 4
> requests_slow_raft{store="1"} 0

Release note: None

Release justification: none, will merge once no justification necessary

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86643)
<!-- Reviewable:end -->

tbg commented:
TODO: I should be able, hopefully, to also add testing for this with one of the existing per-replica circuit breaker tests.

erikgrinaker commented on pkg/kv/kvserver/replica_raft.go:
```diff
@@ -1306,11 +1313,16 @@ func (r *Replica) refreshProposalsLocked(
 				maxSlowProposalDurationRequest = p.Request
-				slowProposalCount++
 			}
+		} else if inflightDuration > time.Minute {
```
Should we have a constant for this? No strong opinion either way.

pavelkalinnikov replied:
Is there already some default for `slowReplicationThreshold` above? Could reuse it here.

erikgrinaker reviewed (approved):
Nice find -- pretty bad that this was missed during review. Can we backport some of this?

pavelkalinnikov commented on pkg/kv/kvserver/replica_metrics.go:
```diff
@@ -85,80 +85,79 @@ func (r *Replica) Metrics(
+	paused             map[roachpb.ReplicaID]struct{}
+}
+
+func calcReplicaMetrics(_ context.Context, d calcReplicaMetricsInput) ReplicaMetrics {
```
Why keeping the unused context?

pavelkalinnikov commented on pkg/kv/kvserver/replica_metrics.go:
```diff
@@ -155,7 +159,7 @@ func calcReplicaMetrics(_ context.Context, d calcReplicaMetricsInput) ReplicaMet
 		m.PausedFollowerCount = int64(len(d.paused))
 	}
-
+	m.SlowRaftProposalCount += d.slowRaftProposalCount
```
Why `+=`? There seems to be no other statement modifying this field here.

pavelkalinnikov commented on pkg/kv/kvserver/replica_metrics.go:
```diff
@@ -86,43 +88,45 @@ func (r *Replica) Metrics(
 	lockTableMetrics := r.concMgr.LockTableMetrics()
 
 	return calcReplicaMetrics(ctx, calcReplicaMetricsInput{
```
Consider creating the `calcReplicaMetricsInput` under the lock above, to dedup most of the variables.

tbg committed:
kvserver: simplify calcReplicaMetrics

Release justification: None needed
Release note: None

tbg committed:
kvserver: simplify calcReplicaMetrics

Rather than populating the struct piecemeal, populate it all at once
which makes it easier to spot omissions.

Release justification: None needed
Release note: None

tbg commented:
CI red due to https://github.com/cockroachdb/cockroach/pull/87684

bors r=erikgrinaker,pavelkalinnikov
